### PR TITLE
Set `templateGroup` for faux requests

### LINKED ISSUE #5552: Customized, renamed templates in template subfolder are not found in RSS feeds
### Affected version(s)

4.9 und 4.13

### Description

Werden in Nachrichten angepasste, umbenannte Templates aus Unterordnern von Templates benutzt, dann werden diese in RSS-Feeds (Exporteinstellung: Ganze Beiträge) nicht gefunden und führen zum Fehler `Could not find template "xyz"` .
Dieser Fehler tritt in Contao 4.9 nicht bei RSCE auf, aber bei umbenannten Core-Templates (z.B.ce_text_mein_template.html5). Laut @ausi wird das in Contao 4.9 für die RSCE abgefangen.
In Contao 4.13 tritt der Fehler dann sowohl für Core-Elemente als auch für die RSCE-Elements auf, weil die Elemente des Feeds im Frontend-Kontext laufen.
Das Problem lässt sich relativ einfach umgehen in dem die entsprechenden Templates in den Rootfolder verschoben werden.
Es sollte diskutiert werden, ob das eine Known Limitation ist oder behoben werden sollte.

Siehe dazu auch https://contao.slack.com/archives/CLTUNH78X/p1670324486656609

### PR DESCRIPTION
**Pull request by fritzmg**
Fixes #5552 (see https://github.com/contao/contao/issues/5552#issuecomment-1339535384)

### CODE REVIEW

File: calendar-bundle/src/Resources/contao/classes/Calendar.php [hunk 1]
```diff
@@ -640,7 +640,19 @@ private function getPageWithDetails($intPageId)
 	{
 		if (!isset(self::$arrPageCache[$intPageId]))
 		{
-			self::$arrPageCache[$intPageId] = PageModel::findWithDetails($intPageId);
+			$objPage = PageModel::findWithDetails($intPageId);
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/contao/classes/Calendar.php]: Can `$objPage` be `null` as well?

Comment (leofeyer) [reply calendar-bundle/src/Resources/contao/classes/Calendar.php]: In case it was unclear where I was going with my question: We should add an early return if `$objPage` is `null` then.

File: calendar-bundle/src/Resources/contao/classes/Calendar.php [hunk 2]
```diff
@@ -651,8 +657,6 @@ private function getPageWithDetails($intPageId)
 			/** @var ThemeModel $objTheme */
 			$objTheme = $objLayout->getRelated('pid');
 			$objPage->templateGroup = $objTheme->templates ?? null;
-
-			self::$arrPageCache[$intPageId] = $objPage;
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/contao/classes/Calendar.php]: Why?

File: news-bundle/src/Resources/contao/classes/News.php [hunk 3]
```diff
@@ -577,19 +577,23 @@ private function getPageWithDetails($intPageId)
...
+			{
+				return null;
+			}
+
 			$objLayout = LayoutModel::findByPk($objPage->layout);
 
 			if (null === $objLayout)
 			{
-				return null;
+				return (self::$arrPageCache[$intPageId] = null);
```
Comment (leofeyer) [inline news-bundle/src/Resources/contao/classes/News.php]: This does not have any effect at all, because we are using `!isset(self::$arrPageCache[$intPageId])` in line 578. Please only add the early return statements and leave the rest as is.

Comment (fritzmg) [reply news-bundle/src/Resources/contao/classes/News.php]: We need to set `self::$arrPageCache[$intPageId]` to `null` here, otherwise `$objLayout = LayoutModel::findByPk($objPage->layout);` will get executed for the same page every time subsequently.

Comment (leofeyer) [reply news-bundle/src/Resources/contao/classes/News.php]: `!isset()` will return `false` if the value is `null`, therefore, as I said above, your changes do not have any effect. 🤷‍♂️

Comment (fritzmg) [reply news-bundle/src/Resources/contao/classes/News.php]: Ah yes, then we need `array_key_exists`.

File: core-bundle/src/Resources/contao/library/Contao/Controller.php [hunk 4]
```diff
@@ -77,9 +77,10 @@ abstract class Controller extends System
 	public static function getTemplate($strTemplate)
 	{
 		$strTemplate = basename($strTemplate);
+		$request = System::getContainer()->get('request_stack')->getCurrentRequest();
 
 		// Check for a theme folder
-		if (\defined('TL_MODE') && TL_MODE == 'FE')
+		if ($request && System::getContainer()->get('contao.routing.scope_matcher')->isFrontendRequest($request))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: This change is unrelated, isn‘t it?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: No, without it the change will not work as `TL_MODE` will be `'BE'`.

File: calendar-bundle/src/Resources/contao/classes/Calendar.php [hunk 5]
```diff
@@ -638,9 +638,25 @@ public function purgeOldFeeds()
 	 */
 	private function getPageWithDetails($intPageId)
 	{
-		if (!isset(self::$arrPageCache[$intPageId]))
+		if (!\array_key_exists($intPageId, self::$arrPageCache))
 		{
-			self::$arrPageCache[$intPageId] = PageModel::findWithDetails($intPageId);
+			$objPage = self::$arrPageCache[$intPageId] = PageModel::findWithDetails($intPageId);
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/contao/classes/Calendar.php:643]: I‘m afraid this does not work. `self::$arrPageCache[$intPageId]` will now store the page model **without** the `templateGroup` parameter, which is only set in line 659.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: calendar-bundle/src/Resources/contao/classes/Calendar.php (hunk 5 above)
Comment (fritzmg) [reply calendar-bundle/src/Resources/contao/classes/Calendar.php:643]: `$objPage` is an object reference, thus any changes to `$objPage` will also be present in `self::$arrPageCache[$intPageId]`. I tested it and it works.

File: core-bundle/src/Resources/contao/library/Contao/Controller.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: But you are not using the `getTemplate()` method in any of the changed code, are you?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: This PR fixes #5552. This issue occurs if you are using a custom template in a theme subfolder, e.g. `foobar/ce_text_foobar.html5` for example and set that template in one of your content elements of one of your news for which you have enabled the news feed. The news feed XML is then generated by Contao within the back end scope, which causes various issues (including #5552). For this reason we introduced a faux request stack in order to simulate a front end request and also simulate a `$GLOBALS['objPage']` object for the respective request.

This PR fixes #5552 by also setting the `templateGroup` on the global page object. However, this change alone will not fix it, as `Controller::getTemplate` (which will be called during rendering of a content element) ignores the `templateGroup` if `TL_MODE` is not `'FE'`. Thus this PR also needs to change this logic to using the ScopeMatcher instead.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: Ah, now I see. 👍

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: > `Controller::getTemplate` (which will be called during rendering of a content element) ignores the `templateGroup` if `TL_MODE` is not `'FE'`.

That seems correct to me, as template groups are only supported in the front end.

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:83]: Yes - hence the change in this Pull Request. Or what is your point?

**Review by ausi — DISMISSED** [review]

File: calendar-bundle/src/Resources/contao/classes/Calendar.php [hunk 6]
```diff
@@ -638,9 +638,25 @@ public function purgeOldFeeds()
...
+
+			if (null === $objPage)
+			{
+				return $objPage;
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/contao/classes/Calendar.php:647]:
```suggestion
				return null;
```

**Review by leofeyer — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.
